File: Calculator/app/src/main/java/com/practical4/calculator/MainActivity.java
package com.practical4.calculator;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {
    private TextView display;
    private String currentInput = "";
    private double memoryValue = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        display = findViewById(R.id.display);
    }

    public void onButtonClick(View view) {
        Button button = (Button) view;
        String buttonText = button.getText().toString();

        switch (buttonText) {
            case "MC":
                memoryValue = 0;
                break;
            case "MR":
                currentInput = String.valueOf(memoryValue);
                break;
            case "MS":
                memoryValue = Double.parseDouble(currentInput);
                break;
            case "=":
                try {
                    currentInput = String.valueOf(eval(currentInput));
                } catch (Exception e) {
                    currentInput = "Error";
                }
                break;
            case "C":
                currentInput = ""; // Clear the display
                break;
            default:
                currentInput += buttonText;
                break;
        }

        display.setText(currentInput);
    }

    // Evaluate the mathematical expression
    private double eval(String expression) {
        return new Object() {
            int pos = -1, ch;

            void nextChar() {
                ch = (++pos < expression.length()) ? expression.charAt(pos) : -1;
            }

            boolean isDigitChar() {
                return (ch >= '0' && ch <= '9') || ch == '.';
            }

            double parse() {
                nextChar();
                double x = parseExpression();
                if (pos < expression.length()) throw new RuntimeException("Unexpected: " + (char)ch);
                return x;
            }

            double parseExpression() {
                double x = parseTerm();
                for (;;) {
                    if (ch == '+') {
                        nextChar();
                        x += parseTerm();
                    } else if (ch == '-') {
                        nextChar();
                        x -= parseTerm();
                    } else {
                        return x;
                    }
                }
            }

            double parseTerm() {
                double x = parseFactor();
                for (;;) {
                    if (ch == '*') {
                        nextChar();
                        x *= parseFactor();
                    } else if (ch == '/') {
                        nextChar();
                        x /= parseFactor();
                    } else {
                        return x;
                    }
                }
            }

            double parseFactor() {
                if (ch == '+') {
                    nextChar();
                    return parseFactor();
                }
                if (ch == '-') {
                    nextChar();
                    return -parseFactor();
                }
                double x;
                int startPos = this.pos;
                if ((ch >= '0' && ch <= '9') || ch == '.') {
                    while (isDigitChar()) nextChar();
                    x = Double.parseDouble(expression.substring(startPos, this.pos));
                } else {
                    throw new RuntimeException("Unexpected: " + (char)ch);
                }
                return x;
            }
        }.parse();
    }
}

File: Prat4/app/src/main/java/com/example/calculatorapp/MainActivity.kt
package com.example.calculatorapp

import android.os.Bundle
import android.text.Editable
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AppCompatActivity


class MainActivity : AppCompatActivity() {
    private var currentValue = ""
    private var storedValue = 0.0
    private var currentOperator = ""
    private var memoryValue = 0.0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Initialize your views here if needed.
    }

    fun onNumberClick(view: View) {
        if (view is Button) {
            val number = view.text.toString()
            currentValue += number
            val display = findViewById<EditText>(R.id.display)
            display.append(number)
        }
    }

    fun onOperatorClick(view: View) {
        if (view is Button) {
            if (currentValue.isNotEmpty()) {
                storedValue = currentValue.toDouble()
                currentOperator = view.tag.toString()
                currentValue = ""
                val display = findViewById<EditText>(R.id.display)
                display.text.clear()
            }
        }
    }

    fun onEqualsClick(view: View) {
        if (currentValue.isNotEmpty()) {
            val num = currentValue.toDouble()
            when (currentOperator) {
                "+" -> storedValue += num
                "-" -> storedValue -= num
                "*" -> storedValue *= num
                "/" -> {
                    if (num != 0.0) {
                        storedValue /= num
                    } else {
                        // Handle division by zero error
                    }
                }
            }
            currentValue = storedValue.toString()
            val display = findViewById<EditText>(R.id.display)
            val editableText = Editable.Factory.getInstance().newEditable(currentValue)
            display.text = editableText
        }
    }

    fun onDecimalClick(view: View) {
        if (view is Button) {
            if (!currentValue.contains(".")) {
                // Ensure there is only one decimal point in the current value
                currentValue += view.text.toString()
                val display = findViewById<EditText>(R.id.display)
                display.append(view.text)
            }
        }
    }


    fun onMemorySaveClick(view: View) {
        if (currentValue.isNotEmpty()) {
            memoryValue = currentValue.toDouble()
        }
    }

    fun onMemoryReadClick(view: View) {
        currentValue = memoryValue.toString()
        val display = findViewById<EditText>(R.id.display)
        val editableText = Editable.Factory.getInstance().newEditable(currentValue)
        display.text = editableText
    }

    fun onMemoryClearClick(view: View) {
        val display = findViewById<EditText>(R.id.display)
        display.text.clear()
    }
}
